Sources/PackageRegistry/SignatureValidation.swift

```diff
@@ -194,13 +195,148 @@ struct SignatureValidation {
+        registry: Registry,
+        package: PackageIdentity.RegistryIdentity,
+        version: Version,
+        toolsVersion: ToolsVersion?,
+        manifestContent: String,
+        configuration: RegistryConfiguration.Security.Signing,
+        timeout: DispatchTimeInterval?,
+        fileSystem: FileSystem,
+        observabilityScope: ObservabilityScope,
+        callbackQueue: DispatchQueue,
+        completion: @escaping (Result<SigningEntity?, Error>) -> Void
+    ) {
+        do {
+            let versionMetadata = try self.versionMetadataProvider(package, version)
+
+            guard let sourceArchiveResource = versionMetadata.sourceArchive else {
+                throw RegistryError.missingSourceArchive
+            }
+            guard sourceArchiveResource.signing?.signatureBase64Encoded != nil else {
+                throw RegistryError.sourceArchiveNotSigned(
+                    registry: registry,
+                    package: package.underlying,
+                    version: version
+                )
+            }
+
+            // source archive is signed, so the manifest must also be signed
+            guard let manifestSignature = try ManifestSignatureParser.parse(utf8String: manifestContent) else {
+                return completion(.failure(RegistryError.manifestNotSigned(
+                    registry: registry,
+                    package: package.underlying,
+                    version: version,
+                    toolsVersion: toolsVersion
+                )))
+            }
+
+            guard let signatureFormat = SignatureFormat(rawValue: manifestSignature.signatureFormat) else {
+                return completion(.failure(RegistryError.unknownSignatureFormat(manifestSignature.signatureFormat)))
+            }
+
+            self.validateSignature(
+                registry: registry,
+                package: package,
+                version: version,
+                signature: manifestSignature.signature,
+                signatureFormat: signatureFormat,
+                content: manifestSignature.contents,
+                configuration: configuration,
+                fileSystem: fileSystem,
+                observabilityScope: observabilityScope,
+                completion: completion
+            )
+        } catch ManifestSignatureParser.Error.malformedManifestSignature {
+            completion(.failure(RegistryError.invalidSignature(reason: "manifest signature is malformed")))
+        } catch RegistryError.sourceArchiveNotSigned(let registry, let package, let version) {
+            observabilityScope.emit(info: "\(package) \(version) from \(registry) is unsigned")
+            guard let onUnsigned = configuration.onUnsigned else {
+                return completion(.failure(RegistryError.missingConfiguration(details: "security.signing.onUnsigned")))
+            }
+
+            let sourceArchiveNotSignedError = RegistryError.sourceArchiveNotSigned(
```

Comment:
iiuc some of these errors are same as for source archive. i think it's fine, but would be nice to have dedicated errors

Reply:
But they ARE related to source archive? Right now we always check if the archive is signed then require manifest signing, so if archive is signed but manifest is not, we throw `.manifestNotSigned` error.  We could make everything driven only by the manifest itself I suppose.

Reply:
it's totally fine as is 👍🏼